Boost callable completions that are not double underscore/default

viniciusmuller opened the pull request:
Lowers the priority of double underscore or default functions when completing functions and captures.

Behavior with this patch: 
![image](https://github.com/lexical-lsp/lexical/assets/59743220/8e2bdffa-ffa4-419a-af3b-1a71ce4e7213)


Closes #183

scohen commented on apps/server/lib/lexical/server/code_intelligence/completion/translations/callable.ex:
```diff
@@ -71,7 +71,13 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.Callable do
-    "#{normalized}/#{arity}"
+    fun = "#{normalized}/#{arity}"
+
+    unless String.starts_with?(name, "__") or name in ["module_info"] do
```
nit: This will fail credo, don't use unless with an else. Favor if.

viniciusmuller commented:
Hey @scohen, I've updated the code to use if instead of unless and also updated a test that was failing since now the completions will be boosted by default (maybe we also could boost to a lower level than 5)

Also, do you think it makes sense to add any other functions to the `["module_info"]` list? I can also try to write a test for this "unprioritized behavior"

scottming commented on apps/server/lib/lexical/server/code_intelligence/completion/translations/callable.ex:
```diff
@@ -71,7 +71,13 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.Callable do
 
   defp sort_text(%_{name: name, arity: arity}) do
     normalized = String.replace(name, "__", "")
-    "#{normalized}/#{arity}"
+    fun = "#{normalized}/#{arity}"
+
+    if String.starts_with?(name, "__") or name in ["module_info"] do
+      fun
+    else
+      Env.boost(fun)
+    end
```
The logic in this section seems convoluted, mainly because the `normalization` and `conditional statements` are far apart and not connected.

Therefore, I prefer some code like this:

```elixir
  defp sort_text(%_{name: name, arity: arity}) do
    if String.starts_with?(name, "__") or name in ["module_info"] do
      "#{normalize(name)}/arity"
    else
      boosted = name |> normalized() |> Enum.boost
      "#{boosted}/arity"
    end
  end

  defp normalize(name) do
    String.replace(name, "__", "")
  end
```

scohen replied:
I think what was there is fine. If you follow the suggestion, you have two places where normalizations happen and the logic to build the sort text is duplicated

scottming replied:
Ok, ignore my suggestion.

scohen commented:
@viniciusmuller please write some tests, and I can't think of any functions that I'd  de-boost right now other than the dunder ones.

viniciusmuller commented on apps/server/test/lexical/server/code_intelligence/completion/translations/function_test.exs:
```diff
@@ -200,5 +200,27 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.FunctionTest d
+      total_completions = length(completions)
+      info_idx = Enum.find_index(completions, &(&1.sort_text == "info/1"))
+
+      assert info_idx > total_completions * 0.80
```
I did some work on this:

```ex
    test "dunder functions have lower completion priority", %{project: project} do
      completions =
        project
        |> complete("GenServer.|")
        |> Enum.sort_by(& &1.sort_text)
        |> Enum.with_index()

      is_dunder = fn {fun, _} -> String.starts_with?(fun.label, "__") end
      dunder_functions = Enum.filter(completions, is_dunder)
      non_dunder = Enum.reject(completions, is_dunder)

      for {_f, idx} <- dunder_functions do
        assert Enum.all?(non_dunder, &smaller_index(&1, idx))
      end
    end

    test "default functions have lower completion priority", %{project: project} do
      completions =
        project
        |> complete("GenServer.|")
        |> Enum.sort_by(& &1.sort_text)
        |> Enum.with_index()

      is_default = fn {fun, _} -> fun.sort_text in ["module_info/0"] end
      default_funs = Enum.filter(completions, is_default)
      other_funs = Enum.reject(completions, is_default)

      for {_f, idx} <- default_funs do
        assert Enum.all?(other_funs, &smaller_index(&1, idx))
      end
    end
  end

  defp smaller_index({_, i}, idx), do: i < idx
```

But I don't like it very much, using this logic, the dunder case fails because the `module_info` function is not a dunder function and gets after it, and I don't think it would be ideal to filter it out in the dunder test case.

Maybe a simpler solution would be: get all the dunder and default functions and assert they do not start with a booster character, what do you think @scohen?

scohen replied:
> Maybe a simpler solution would be: get all the dunder and default functions and assert they do not start with a booster character, what do you think

it might even be easier than that. Make a function called `boosted?` that checks for a boost character. Then it's 
```elixir
completions = 
project
|> complete("GenServer.|")
|> Enum.sort_by(& &1.sort_text)

{dunder_completions, non_dunder_completions} = Enum.split_with(completions, &String.starts_with?(&1.label, "__"))

for completion <- dunder_completions do 
  refute boosted?(completion)
end

for completion <- non_dunder_functions do 
  assert boosted?(completion)
end
```
